opentelemetry/src/context.rs

```diff
@@ -381,6 +387,87 @@ impl Hasher for IdHasher {
     }
 }
 
+/// A stack for keeping track of the [`Context`] instances that have been attached
+/// to a thread.
+///
+/// The stack allows for popping of contexts by position, which is used to do out
+/// of order dropping of [`ContextGuard`] instances. Only when the top of the
+/// stack is popped, the topmost [`Context`] is actually restored.
+///
+/// The stack relies on the fact that it is thread local and that the
+/// [`ContextGuard`] instances that are constructed using it can't be shared with
+/// other threads.
+struct ContextStack {
+    /// This is the current [`Context`] that is active on this thread, and the top
+    /// of the [`ContextStack`]. It is always present, and if the `stack` is empty
+    /// it's an empty [`Context`].
+    ///
+    /// Having this here allows for fast access to the current [`Context`].
+    current_cx: Context,
+    /// A `stack` of the other contexts that have been attached to the thread.
+    stack: Vec<Option<Context>>,
+    /// Ensure this type is !Send as it relies on thread locals
+    _marker: PhantomData<*const ()>,
+}
+
+impl ContextStack {
+    #[inline(always)]
+    fn push(&mut self, cx: Context) -> usize {
+        // The next id is the length of the `stack`, plus one since we have the
+        // top of the [`ContextStack`] as the `current_cx`.
+        let next_id = self.stack.len() + 1;
+        let current_cx = std::mem::replace(&mut self.current_cx, cx);
+        self.stack.push(Some(current_cx));
+        next_id
+    }
+
+    #[inline(always)]
+    fn pop_id(&mut self, pos: usize) {
+        if pos == 0 {
+            // The empty context is always at the bottom of the [`ContextStack`]
+            // and cannot be popped, so do nothing.
+            return;
+        }
+        let len = self.stack.len();
+        // Are we at the top of the [`ContextStack`]?
+        if pos == len {
+            // Shrink the stack if possible to clear out any out of order pops.
+            while let Some(None) = self.stack.last() {
+                _ = self.stack.pop();
+            }
+            // Restore the previous context. This will always happen since the
+            // empty context is always at the bottom of the stack if the
+            // [`ContextStack`] is not empty.
+            if let Some(Some(next_cx)) = self.stack.pop() {
+                self.current_cx = next_cx;
+            }
+        } else {
+            // This is an out of order pop.
+            if pos >= len {
+                // This is an invalid id, ignore it.
+                return;
+            }
+            // Clear out the entry at the given id.
+            _ = self.stack[pos].take();
+        }
+    }
+
+    #[inline(always)]
+    fn map_current_cx<T>(&self, f: impl FnOnce(&Context) -> T) -> T {
+        f(&self.current_cx)
+    }
+}
+
+impl Default for ContextStack {
+    fn default() -> Self {
+        ContextStack {
+            current_cx: Context::default(),
+            stack: Vec::with_capacity(64),
```

Comment:
do we need to pre-allocate? It can just grow as needed, right? And once allocated, it won't shrink allocated capacity...

Reply:
We don't need to preallocate, but at the same time, this is a thread local stack, and I think that preparing for a _reasonable_ number of spans/contexts is ok.

As for shrinking, I think that we should add some heuristics for doing that when the gap between that capacity and top of the stack becomes large. There is always the risk that the stack will inflate/deflate in a ping-pong fashion, so we need to decide what is more preferable.

Reply:
Got it. Context of 64 depth may also be a lot. Maybe stick with a lower number like 8 or so. (don't have any data to support any number, so just picked 8 to be conservative!)

The shrink part is not going to be common right? If user does attach/detach in the expected order, there should not be any gaps/holes

Reply:
Yes, the shrink part should not be common, or if the depth of spans balloons up, and then down, it seems likely that it will do that again.